### ISSUE TITLE
Editors: revisit `forceFile`, `forceUntitled`, `setForceOpenAsText` and `setForceOpenAsBinary`

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#128209]
There are a few things that smell like they are unsupported in the nice new untyped editor world:
* [`IBaseResourceEditorInput.forceFile`](https://github.com/microsoft/vscode/blob/45298b83eab670f2531b08791b8c3fe8c1b63cd1/src/vs/platform/editor/common/editor.ts#L61-L61)
* [`IBaseResourceEditorInput.forceUntitled`](https://github.com/microsoft/vscode/blob/45298b83eab670f2531b08791b8c3fe8c1b63cd1/src/vs/platform/editor/common/editor.ts#L70-L70)
* [`FileEditorInput.setForceOpenAsText`](https://github.com/microsoft/vscode/blob/74264ceab580be39f5091068453f53ae2296d9c2/src/vs/workbench/contrib/files/browser/editors/fileEditorInput.ts#L265-L265)
* [`FileEditorInput.setForceOpenAsBinary`](https://github.com/microsoft/vscode/blob/74264ceab580be39f5091068453f53ae2296d9c2/src/vs/workbench/contrib/files/browser/editors/fileEditorInput.ts#L269-L269)

**Comment by lramos15** [comment]
We do handle `forceUntitled` https://github.com/microsoft/vscode/blob/45298b83eab670f2531b08791b8c3fe8c1b63cd1/src/vs/workbench/common/editor.ts#L339-L339

`forceFile` support is interesting, do we always consider it a text editor?

`setForceOpenAsText` and `setForceOpenAsBinary` should probably just become options in `IEditorOptions` and then the factory functions will know what to do

**Comment by bpasero** [comment]
* `forceFile`: I think this is something to nuke
* `forceUntitled`: this should maybe change to `isUntitledWithAssociatedResource` or similar
* `setForceOpenAsText` & `setForceOpenAsBinary` are very specific for file editors because by default we refuse to open any file as text that is binary. I don't think we want to expose this to untyped inputs or other editors

**Comment by lramos15** [comment]
> * `forceFile`: I think this is something to nuke

This is pretty much the equivalent of `override: DEFAULT_EDITOR_ASSOCIATION.id` and could be replaced with that

> * `setForceOpenAsText` & `setForceOpenAsBinary` are very specific for file editors because by default we refuse to open any file as text that is binary. I don't think we want to expose this to untyped inputs or other editors

I'm just wondering how can we handle this in the resolver / open flow without having to resolve a typed input outside the editor service and then call the correct binary / text method

**Comment by bpasero** [comment]
Yeah good points, I can maybe start to look into this when I am back, didn't mean we have to tackle this already now 👍

**Comment by bpasero** [comment]
@lramos15 spend some time today to look into it, here are my findings:
* `setForceOpenAsText` and `setForceOpenAsBinary` are used exclusively from editors that are associated with file editor inputs, so I actually no longer believe this is debt to clean up because we do not leak this outside of that universe
* `forceUntitled` is actually useful and we cannot drop it: we use it to enforce to create untitled editor input in the resolver even if the provided resource is not `untitled`. so it enables clients to resolve an untitled editor with associated resource and I suggest to keep it `forceUntitled`
* `forceFile` is problematic: this is actually NOT the same as using `override: DEFAULT_EDITOR_ASSOCIATION.id` because it is being used as a hint to decide whether a file system provider check is needed or not:

https://github.com/microsoft/vscode/blob/941c07bb6c141b43c193bcccad60acbd8e91aee8/src/vs/workbench/services/editor/browser/editorService.ts#L964-L964

I think I had to do this because:
* file editors are possibly created so early that extensions did not have a chance to activate and as such the check for `canHandleResource` might fail even though the extension exists
* because `createEditorInput` is not `async` we cannot trigger our usual `fileService.activateProvider` call because that returns a `Promise`

I now deprecated `forceFile` but have no good answer how to solve this better. I think one possible solution would be to merge `FileEditorInput` and `TextResourceEditorInput` into one so that the decision about this can be deferred to a later point in time. Both inputs will end up using mostly the same editor widget and model but the latter is readonly.

**Comment by bpasero** [comment]
@mjbvz @lramos15 can you please chime in for the use of `forceFile` here:

https://github.com/microsoft/vscode/blob/5f3d88ad79fbdad08778620a639b6113e0e677dd/src/vs/workbench/contrib/customEditor/browser/customEditorInput.ts#L41-L41

and 

https://github.com/microsoft/vscode/blob/5f3d88ad79fbdad08778620a639b6113e0e677dd/src/vs/workbench/contrib/customEditor/browser/customEditorInput.ts#L343-L343

**Comment by lramos15** [comment]
> * `setForceOpenAsText` and `setForceOpenAsBinary` are used exclusively from editors that are associated with file editor inputs, so I actually no longer believe this is debt to clean up because we do not leak this outside of that universe

Understandable, this is just difficult to handle correctly in those particular cases because we need to resolve the input then set the options on it afterwards by calling this function. It's very different than every other input type

> * `forceUntitled` is actually useful and we cannot drop it: we use it to enforce to create untitled editor input in the resolver even if the provided resource is not `untitled`. so it enables clients to resolve an untitled editor with associated resource and I suggest to keep it `forceUntitled`

The resolver service already respects forceUntitled so I think we're ok here

> I now deprecated `forceFile` but have no good answer how to solve this better. I think one possible solution would be to merge `FileEditorInput` and `TextResourceEditorInput` into one so that the decision about this can be deferred to a later point in time. Both inputs will end up using mostly the same editor widget and model but the latter is readonly.

Might not be understanding the problem wrong but yes editor input creation is synchronous but we resolve which input to create in an async fashion so could we make a request to the fileService provider before deciding on the input?



> @mjbvz @lramos15 can you please chime in for the use of `forceFile` here:

I think I just followed what Matt was doing and it seemed like we used it here because we wanted a text editor. forceFile wasn't well documented, wasn't aware that was what it was used for

**Comment by bpasero** [comment]
@lramos15 

> Understandable, this is just difficult to handle correctly in those particular cases because we need to resolve the input then set the options on it afterwards by calling this function. It's very different than every other input type

I am not so worried about these methods and would not think that we win anything but adding support for it to untyped inputs because this would need implementation and buy-in from all editors that contribute to the resolver which is unrealistic imho.

Especially since the only callers of these methods are the very editors that already show the typed input to begin with.

> Might not be understanding the problem wrong but yes editor input creation is synchronous but we resolve which input to create in an async fashion so could we make a request to the fileService provider before deciding on the input?

I think the actual challenge here is that a lot of editor inputs that are text based share the same editor id (`default`) but that is not sufficient to distinguish e.g. a file editor input from a text resource input. What I actually need is a way to set the `override` specifically to one of the 2 editors and then we could drop the `forceFile`.

> I think I just followed what Matt was doing and it seemed like we used it here because we wanted a text editor. 

Ok perfect then I will change it to use the `override` instead.

**Comment by lramos15** [comment]
> What I actually need is a way to set the `override` specifically to one of the 2 editors and then we could drop the `forceFile`.

We could always split from the default and give them two separate ids, although I'm not sure the implications of that as I often use `default` to specify "give me a text editor". However, this seems challenging because those inputs are both text based. What are the differences between the two and is there one that would normally be "default" and the others are special cases?

**Comment by bpasero** [comment]
The file editor input specifically requires a registered file system provider (to the `IFileService`). The other one can work provided there is a text document content provider and is readonly. 

I will continue to look into this further but may decide to merge this into https://github.com/microsoft/vscode/issues/126960 because once we no longer call `createEditorInput` from the outside, we can invent some internal way of resolving file editor input vs. text resource input that does not bleed into our APIs.

**Comment by bpasero** [comment]
Ended up allowing specifically to `createEditorInput: IFileEditorInput` and only use it in very few places. We can further clean this up once https://github.com/microsoft/vscode/issues/126960 lands.
